### DISCUSSION TITLE
Water Consumption Prediction Using Heat Water Tank Temperatures

### DISCUSSION THREAD

**Discussion by sachaalidadi** [master-csmi/csmi#18] [Project Portfolio]
# Introduction
To provide thermal comfort and energy savings to their customers, De Dietrich has recognized the need to measure water consumption accurately. This facilitates adapting hot water production according to the customer's specific needs, thus promoting efficient energy usage. Traditionally, this tracking of water volume consumed by a hot water tank is achieved through flow meters. However, these high-quality flow meters carry a substantial cost, both financially and environmentally. They are not only expensive to acquire but also contribute significantly to the carbon footprint due to their manufacturing processes. Also, new regulations forbid the company to sell devices with a high carbon footprint. An alternative to these devices is thermometers, which are considerably less expensive and have a significantly lower carbon footprint.

# Methods
A solution to this problem is to use the temperatures of the water at different layers of the water tank. Because the more you consumed water the lower the temperature will be at the bottom.

<img width="629" alt="Stratification" src="https://github.com/master-csmi/csmi/assets/114726985/fbcd58af-42e2-448e-a16c-5a274e27a642">

Throughout the project, the approach chosen is a two-model approach. The first model is a classification model that predicts if there is a drawing at an instant t. The second model is a regression model that predicts the volume of water consumed by taking the temperature and the output of the first model.

The machine learning algorithms used are the following:

- Linear Regression

This method consists to find the best a and b such that a*x + b is very close to y, where y is the target data (Volume consumed) and x is the data used (temperatures).

- Random Forest

A random forest is a machine-learning algorithm that combines the predictions of multiple decision trees to make more accurate and robust predictions.

Each decision tree is trained with the CART algorithm (Classification And Regression Tree). The idea is actually quite simple: the algorithm first divides the `train` set into two subsets (=nodes) using a single characteristic $k$ and a threshold $t_k$ (for example, "temperature at layer 1 $\leq$ 30°C").

The couple $(k,t_k)$ is chosen according to which couple produces the lowest cost function. This cost function is computed like so:

$$
J(k,t_k)= \frac {m_{\text{left}} } {m} \,  G_{\text{left}} +  \frac {m_{\text{right}} } {m} \,  G_{\text{right}}
$$

where:

$m_{\text{left/right}}$ is the number of instances in the left/right node
$G_{\text{left/right}}$ is the Gini index of the left/right node

- Adaboost

It is an ensemble learning and boosting algorithm. The models are trained sequentially and each one tries to improve the results of its predecessor.

- Gradient Boosting

It is an ensemble learning and boosting algorithm. The models are trained sequentially and each one tries to improve the results of its predecessor.

The model is optimized like so:

- We compute the prediction of the model $F_{m-1}$ on the training set.

- We substract the prediction of the model $F_{m-1}$ from the target data.

- We train the model $F_m$ on the new target data.

To measure the performance of the models, the following metrics are used:

In the following formulas, $y_i$ is the target data, and $\hat{y}_i$ is the predicted data.

- RMSE (Root Mean Square Error)

$$ RMSE = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2} $$

- R2 score

$$ R^2 = 1 - \frac{\sum_{i=1}^{n} (y_i - \hat{y}_i)^2}{\sum_{i=1}^{n} (y_i - \bar{y})^2} $$

- F1 score (This metric is used to optimize the classification model)

$$ F_1 = 2 \cdot \frac{precision \cdot recall}{precision + recall} $$

Where:

$$ precision = \frac{TP}{TP + FP} $$

$$ recall = \frac{TP}{TP + FN} $$

$TP$ is the number of true positives, $FP$ is the number of false positives and $FN$ is the number of false negatives.

A lot of time has been spent optimizing the first model. Because the f1 score was deficient no matter the model used.

Here are the results of the different models:

|   Model   |   RMSE |   R2 score |   F1 score |
|---    |:-:    |:-:    |:-:    |
|   RandomForest   |   0.14   |   0.944 |   0.93 |
|   GradientBoosting   |  0.15   |   0.940 |   0.49 |
|   AdaBoost   |   0.26 | 0.813 |   0.20 |

We can now take a look at the violin plot of the best model.

If you're not familiar, a Box Plot is a statistical representation of the data distribution. The box represents the interquartile range (25th to 75th percentiles), the white dot inside the box is the median, and the whiskers (lines extending from the box) represent the variability outside the upper and lower quartiles, giving us a sense of data spread and outliers.
On the other hand, a Violin Plot gives a more detailed representation. It combines a box plot and a density plot. The wider sections of the 'violin' represent a higher probability that members of the population will take on the given value; the skinnier sections represent a lower probability.

In this violin plot, we separated the drawings into five categories:

- No drawing
- Low drawing
- Medium drawing
- High drawing
- Huge drawing

![violin_plot_classification_30_10](https://github.com/master-csmi/csmi/assets/114726985/ec2a63a3-090a-43b9-be32-602d6375744f)

Let's take a look closer at each category.

![violin_plot_classifiaction_30_10_none](https://github.com/master-csmi/csmi/assets/114726985/635bfa73-4f8d-412e-bd0d-09b43d133d0e)

![violin_plot_classifiaction_30_10_low](https://github.com/master-csmi/csmi/assets/114726985/14d468be-5a36-4daf-9c9b-c0d162f47057)

![violin_plot_classifiaction_30_10_medium](https://github.com/master-csmi/csmi/assets/114726985/87ee903a-3c64-41ca-8874-596c91c32542)

![violin_plot_classification_30_10_high](https://github.com/master-csmi/csmi/assets/114726985/78a198aa-0ee7-4f5d-b5fe-58394cd9e348)

![violin_plot_classification_30_10_huge](https://github.com/master-csmi/csmi/assets/114726985/685578db-d9a7-460d-8f7d-f3cd8ae23615)

# Conclusion

Even though, the metrics showed promising results for the classification model, the violin plot shows that the model is not able to detect some drawings, especially the low and medium drawings. This is a problem because some days, 100% of the water consumed comes from low and medium drawings.

For future projects, here are the things that could be done:

- Get more data to train the model and reduce overfitting
- Try deep learning methods such as LSTM
- Try a different approach than the two-model approach
- Optimize the hyperparameters of the models
